packages/service-library/src/servicelib/fastapi/tracing.py

```diff
@@ -127,3 +136,66 @@ def setup_tracing(
 
 def setup_httpx_client_tracing(client: AsyncClient | Client):
     HTTPXClientInstrumentor.instrument_client(client)
+
+
+def _opentelemetry_function_span(func: Callable):
+    """Decorator that wraps a function call in an OpenTelemetry span."""
+    tracer = trace.get_tracer(__name__)
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        with tracer.start_as_current_span(f"{func.__module__}.{func.__name__}"):
+            return func(*args, **kwargs)
+
+    @wraps(func)
+    async def async_wrapper(*args, **kwargs):
+        with tracer.start_as_current_span(f"{func.__module__}.{func.__name__}"):
+            return await func(*args, **kwargs)
+
+    if inspect.iscoroutinefunction(func):
+        return async_wrapper
+    else:
+        return wrapper
+
+
+def _opentelemetry_method_span(cls):
+    for name, value in cls.__dict__.items():
+        if callable(value) and not name.startswith("_"):
+            setattr(cls, name, _opentelemetry_function_span(value))
+    return cls
+
+
+class _AddTracingSpansLoader(Loader):
+    def __init__(self, loader: Loader):
+        self.loader = loader
+
+    def exec_module(self, module: ModuleType):
+        self.loader.exec_module(module)
+        for name, func in inspect.getmembers(module, inspect.isfunction):
+            if name in module.__dict__:
+                setattr(module, name, _opentelemetry_function_span(func))
+        for name, cls in inspect.getmembers(module, inspect.isclass):
+            if name in module.__dict__ and cls.__module__ == module.__name__:
+                setattr(module, name, _opentelemetry_method_span(cls))
+
+
+class _AddTracingSpansFinder(MetaPathFinder):
+    def find_spec(
+        self,
+        fullname: str,
+        path: Sequence[str] | None,
+        target: ModuleType | None = None,
+    ) -> ModuleSpec | None:
+        if fullname.startswith("simcore_service"):
+            spec = importlib.machinery.PathFinder.find_spec(
+                fullname=fullname, path=path
+            )
+            if spec and spec.loader:
+                spec.loader = _AddTracingSpansLoader(spec.loader)
+                return spec
+
+        return None
+
+
+def setup_tracing_spans_for_simcore_service_functions():
```

Comment:
why dont you pass the name of the module instead of relying on things that were imported?

If i undestand it right, yoy want to decorate all members and functions of a given module so the steps would be

1. import all modules in a package
2. inspect all members and functions
3. apply decorators above

for 1 and 2 see what we do  with [`walk_model_examples_in_package`](https://github.com/itisfoundation/osparc-simcore/blob/5732a12e07e63d5ce55010ede9b9ab543bb9b278/packages/pytest-simcore/src/pytest_simcore/pydantic_models.py). I guess is similar to the MetaPathFinder but higher level since  uses built-in `pkgutil.walk_packages`